kdc/kerberos5.c

```diff
@@ -472,31 +511,39 @@ pa_pkinit_validate(astgs_request_t r, const PA_DATA *pa)
 	ret = KRB5KRB_AP_ERR_BAD_INTEGRITY;
 	_kdc_r_log(r, 4, "Failed to decode PKINIT PA-DATA -- %s",
 		   r->cname);
+	auth_state->auth_status = HDB_AUTH_PKINIT_FAILURE;
 	goto out;
     }
-    
+
     ret = _kdc_pk_check_client(r, pkp, &client_cert);
     if (ret) {
 	_kdc_set_e_text(r, "PKINIT certificate not allowed to "
 			"impersonate principal");
+	auth_state->auth_status = HDB_AUTH_PKINIT_FAILURE;
 	goto out;
     }
+    auth_state->auth_details = client_cert;
+    auth_state->free_ptr = client_cert;
```

Comment:
`free_ptr` feels a little unidiomatic Heimdal. Perhaps an integer indicating whether it should be freed would be better.

Reply:
Looking at @josephsutton1's new patch, now I see why @metze-samba did it this way.  The distinct pointer is needed so that a const pointer is not passed to free().  This is allowed in Heimdal (perhaps with warnings) but Samba's stricter compile settings won't allow that.  So by keeping two copies of the pointer, it works when set to a real constant string (`"foo"`) and a `malloc()` based string.

So sorry, we probably do need that :-(

Reply:
I'm guilty of the `freeme` style...  I think this is OK.